### PR TITLE
adjust exporter to committee consensus

### PR DESCRIPTION
**Pull request by guym-blox**
This pull request enhances the exporter node to support committee consensus in the SSV system. Previously, the exporter tracked participation using decided messages from other nodes for each validator. With the shift to committee consensus, the exporter now processes consensus data at the committee level, including the switch from decided messages to partial signature post consensus messages.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -40,7 +40,8 @@ Build stage Docker image:
     - docker tag $IMAGE_NAME:$CI_COMMIT_SHA $DOCKER_REPO_INFRA_STAGE:$CI_COMMIT_SHA
     - $DOCKER_LOGIN_TO_INFRA_STAGE_REPO && docker push $DOCKER_REPO_INFRA_STAGE:$CI_COMMIT_SHA
   only:
-    - alan/no-fork
+    - gm/exporter-alan-merge
```
Comment (y0sher) [inline .gitlab-ci.yml:43]: revert

File: cli/operator/node.go [hunk 2]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"encoding/base64"
 	"fmt"
+	"github.com/ssvlabs/ssv/exporter/exporter_message"
```
Comment (y0sher) [inline cli/operator/node.go:7]: format the file and imports

File: cli/operator/node.go [hunk 3]
```diff
@@ -272,18 +273,20 @@ var StartNodeCmd = &cobra.Command{
 			ws := exporterapi.NewWsServer(cmd.Context(), nil, http.NewServeMux(), cfg.WithPing)
 			cfg.SSVOptions.WS = ws
 			cfg.SSVOptions.WsAPIPort = cfg.WsAPIPort
-			cfg.SSVOptions.ValidatorOptions.NewDecidedHandler = decided.NewStreamPublisher(logger, ws)
+			cfg.SSVOptions.ValidatorOptions.NewDecidedHandler = decided.NewStreamPublisher(logger, ws, cfg.SSVOptions.ValidatorOptions.UseNewExporterAPI)
```
Comment (y0sher) [inline cli/operator/node.go:276]: why do we need this flag? old API probably not working for committee conesnsus.

File: config/exporter.yaml [hunk 4]
```diff
@@ -0,0 +1,3 @@
...
...WFZQMVNUM1RERGhSa0d2eUtUVExVWUpidFp2bjQ4MQp3MWppajBJK2tEZjdLb0RnVTRDUnN6N0lCdzRwVEEvVWphMFkrV294ODkxTlVzVFg5YktXOHJWZXMxNDc2TzNHCnZzaVpqT1RVMWhHNmlDdCt4ZXV4S1VKQlVnSUFIaVloK0JZQXJtSEwvam5waDZ5UXQ5ZUFQa1VDZ1lFQThBV0sKVGZWeTY0N1Y2Wmw0SGtpYzA0OTA3dzhGb1B1b0pKbkRPaTBPT1FacWdzbHpmbytpb1pNWFpib2hlWVlrMkRWSgpLdXZ3bXlSYnp3clZSMHRsMUtZMUpCUlBPUnZJZ05OcUd4U1VPOVNHdGp5WDJkeGw5NnFyeXpYRWZBZDlPbDZ5CkRVbUhVazNPU0ptUUFVZUlNQWVja3dCbHJQNkFGUHF6YkRseERVa0NnWUFtVzNodUwwQkFXMWZ0Mm9ldGtzdFUKN1ppS29HWUNTNVduQjE3YUFxWGI4U2FMTUhld042OG1iRnZqNGRDK1N2NHRNLzdSSldBbFp6a291U3BiS1F2egpqZDBEdXpTb0xXZnlmUlRHM1RFUlBaQURvVDh2bm5wLytpdHBET1FoUlNqZm9PWWFSTS9tbnBhRnVnSXpvNGp4Clk0MjFISFNOcG1Oc0dwTHhITmpVMFFLQmdRRGNTZWx1dFlnOHc4SjJmR0V2UWl1VlRBUGhhS01jNWZ1alkvQTMKRTZVTEYyVm9GZVpIVVNMTk1rZjJHSWVpbGtoNG15ZEZHbmlnbXhVY0IvQkpWOXprOTRxc1MydU9tWDVESmtZcgpCbzhwL2x1SkpFR21OVTdzcVJxMURlc0Fmb3NqcUVlNFJEVExpQmQ4c0Z2bm9MUlFpSnZlbXkyMlFYcGlSdG1uCnlXeUY2UUtCZ1FDVHNCemRZOVRuWTZXMkgyTnBtVGg0c1VqV3BIRUFXa01hUnJDMkp1b0p3Mkp3TExjRGFFVEMKaGhia1BmY2pnb2draWxoUmRYaktWM29IVkxsN1d6Rlp4OHRqU29LRDU0eVRtV1JoL2JsWDBOZFFmQjhUcnNoWQpONkhqV2RRLys1NFJkOEdlTHRMVkF5b1B1eTRnalNPUnJReWVKWC9US01CV1hqcVVyL0pyN1E9PQotLS0tLUVORCBSU0EgUFJJVkFURSBLRVktLS0tLQo=
```
Comment (y0sher) [inline config/exporter.yaml:3]: is this a real key? remove it if so

File: eth/eventhandler/handlers.go [hunk 5]
```diff
@@ -5,6 +5,7 @@ import (
 	"encoding/hex"
 	"errors"
 	"fmt"
+	"github.com/ssvlabs/ssv/exporter/exporter_message"
```
Comment (y0sher) [inline eth/eventhandler/handlers.go:8]: format the file gofmt goimports

File: exporter/api/decided/stream.go [hunk 6]
```diff
@@ -18,26 +16,20 @@ import (
 
 // NewStreamPublisher handles incoming newly decided messages.
 // it forward messages to websocket stream, where messages are cached (1m TTL) to avoid flooding
-func NewStreamPublisher(logger *zap.Logger, ws api.WebSocketServer) controller.NewDecidedHandler {
+// TODO: remove useNewAPI parameter when explorer migrates to the new API
```
Comment (y0sher) [inline exporter/api/decided/stream.go:19]: whats stopping you from removing this?

File: exporter/api/msg.go [hunk 7]
```diff
@@ -18,91 +22,74 @@ type Message struct {
...
 		Type: TypeDecided,
 		Filter: MessageFilter{
-			PublicKey: hex.EncodeToString(dutyExecutorID),
-			From:      uint64(decMsg.Height),
-			To:        uint64(decMsg2.Height),
-			Role:      role.String(),
+			PublicKey: hex.EncodeToString(pkv),
+			From:      uint64(msgs[0].Slot),
+			To:        uint64(msgs[len(msgs)-1].Slot),
+			Role:      msgs[0].Identifier.GetRoleType().String(),
```
Comment (y0sher) [inline exporter/api/msg.go:60]: are you sure the messages are sorted by the slot? maybe its best to find the from/to slots in a different way ( loop thorugh and save max and min slot seen )

File: .k8/hetzner-stage/ssv-exporter-holesky.yml [hunk 8]
```diff
@@ -118,6 +118,8 @@ spec:
               value: "16073"
             - name: ENABLE_PROFILE
               value: "true"
+            - name: USE_NEW_EXPORTER_API
+              value: "true"
```
Comment (y0sher) [inline .k8/hetzner-stage/ssv-exporter-holesky.yml:122]: do we stlil need this? in alan you must use the new api right?

File: exporter/api/query_handlers.go [hunk 9]
```diff
@@ -98,3 +44,56 @@ func HandleUnknownQuery(logger *zap.Logger, nm *NetworkMessage) {
...
+	if err != nil {
+		logger.Warn("failed to parse role", zap.Error(err))
+		res.Data = []string{"role doesn't exist"}
+		nm.Msg = res
+		return
+	}
+	runnerRole := exporter_message.RunnerRole(beaconRole)
+	roleStorage := qbftStorage.Get(runnerRole)
+	if roleStorage == nil {
+		logger.Warn("role storage doesn't exist", fields.Role(spectypes.RunnerRole(runnerRole)))
```
Comment (y0sher) [inline exporter/api/query_handlers.go:77]: `spectypes.RunnerRole(runnerRole)` this cast might error because we don't have all `exporter_message.RunnerRole`'s in `spectypes.RunnerRole`

File: exporter/exporter_message/message.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
+package exporter_message
```
Comment (y0sher) [inline exporter/exporter_message/message.go:1]: package name is a bit too verbose, `exporter`/`exporter_message`, its best to call it `message` or `messages`, maybe even `types`, or `convert`.
reapting the expoter is just redundant.

File: ibft/storage/store_test.go [hunk 11]
```diff
@@ -178,3 +179,24 @@ func newTestIbftStorage(logger *zap.Logger, prefix string) (qbftstorage.QBFTStor
...
+func TestEncodeDecodeOperators(t *testing.T) {
+	testCases := []struct {
+		input   []uint64
+		encoded []byte
+	}{
+		{[]uint64{0x0123456789ABCDEF, 0xFEDCBA9876543210}, []byte{0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF, 0xFE, 0xDC, 0xBA, 0x98, 0x76, 0x54, 0x32, 0x10}},
+		{[]uint64{0, 1, 2, 3}, []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 3}},
+		{[]uint64{}, []byte{}},
+	}
+
```
Comment (y0sher) [inline ibft/storage/store_test.go:192]: operators list size can only be 4/7/13, you can reflect this in test and in the functions.

File: ibft/storage/stores.go [hunk 12]
```diff
@@ -2,24 +2,24 @@ package storage
...
+	"github.com/ssvlabs/ssv/exporter/exporter_message"
 
 	qbftstorage "github.com/ssvlabs/ssv/protocol/v2/qbft/storage"
 	"github.com/ssvlabs/ssv/storage/basedb"
 )
 
 // QBFTStores wraps sync map with cast functions to qbft store
 type QBFTStores struct {
-	m *hashmap.Map[spectypes.RunnerRole, qbftstorage.QBFTStore]
+	m *hashmap.Map[exporter_message.RunnerRole, qbftstorage.QBFTStore]
```
Comment (y0sher) [inline ibft/storage/stores.go:13]: maybe `exporter_message.RunnerRole` types and functions and better fitting to this package since they are used mostly here? or when dealing with the storage here.

File: protocol/v2/qbft/instance/instance.go [hunk 13]
```diff
@@ -71,6 +73,7 @@ func (i *Instance) Start(logger *zap.Logger, value []byte, height specqbft.Heigh
 		i.bumpToRound(specqbft.FirstRound)
 		i.State.Height = height
 		i.metrics.StartStage()
+		i.started = time.Now()
```
Comment (y0sher) [inline protocol/v2/qbft/instance/instance.go:76]: why is this part of this PR? if not related please move to other PR

File: protocol/v2/qbft/instance/prepare.go [hunk 14]
```diff
@@ -53,7 +54,9 @@ func (i *Instance) uponPrepare(logger *zap.Logger, signedPrepare *spectypes.Sign
 	logger.Debug("🎯 got prepare quorum",
 		fields.Round(i.State.Round),
 		zap.Any("prepare-signers", allSigners(prepareMsgContainer.MessagesForRound(i.State.Round))),
-		fields.Root(proposedRoot))
+		fields.Root(proposedRoot),
+		fields.QuorumTime(time.Since(i.started)),
```
Comment (y0sher) [inline protocol/v2/qbft/instance/prepare.go:58]: is this part of the exporter PR?

File: protocol/v2/ssv/runner/metrics/metrics.go [hunk 15]
```diff
@@ -64,6 +65,11 @@ var (
 		Name: "ssv_instances_decided",
 		Help: "Number of decided QBFT instances",
 	}, []string{"role"})
+	metricsBeaconDataDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
```
Comment (y0sher) [inline protocol/v2/ssv/runner/metrics/metrics.go:68]: why is this in this PR?

File: protocol/v2/ssv/runner/aggregator.go [hunk 16]
```diff
@@ -216,18 +213,31 @@ func (r *AggregatorRunner) ProcessPostConsensus(logger *zap.Logger, signedMsg *s
...
+		start := time.Now()
+		endSubmission := r.metrics.StartBeaconSubmission()
 
+		logger = logger.With(
+			zap.Uint64s("signers", getPostConsensusSigners(r.GetState(), root)),
+			fields.PreConsensusTime(r.metrics.GetPreConsensusTime()),
+			fields.ConsensusTime(r.metrics.GetConsensusTime()),
+			fields.PostConsensusTime(r.metrics.GetPostConsensusTime()),
+			zap.String("block_root", hex.EncodeToString(msg.Message.Aggregate.Data.BeaconBlockRoot[:])),
+		)
```
Comment (y0sher) [inline protocol/v2/ssv/runner/aggregator.go:225]: not related

File: protocol/v2/ssv/runner/proposer.go [hunk 17]
```diff
@@ -106,30 +103,36 @@ func (r *ProposerRunner) ProcessPreConsensus(logger *zap.Logger, signedMsg *spec
...
 		return errors.Wrap(err, "got pre-consensus quorum but it has invalid signatures")
 	}
-
+	r.metrics.EndPreConsensus()
 	logger.Debug("🧩 reconstructed partial RANDAO signatures",
-		zap.Uint64s("signers", getPreConsensusSigners(r.GetState(), root)))
+		zap.Uint64s("signers", getPreConsensusSigners(r.GetState(), root)),
+		fields.PreConsensusTime(r.metrics.GetPreConsensusTime()))
 
+	r.metrics.StartBeaconData()
```
Comment (y0sher) [inline protocol/v2/ssv/runner/proposer.go:111]: not relaetd to this PR

File: protocol/v2/ssv/runner/sync_committee.go [hunk 18]
```diff
@@ -1,6 +1,5 @@
 package runner
 
-//
```
Comment (y0sher) [inline protocol/v2/ssv/runner/sync_committee.go:3]: ??

File: protocol/v2/ssv/runner/sync_committee_aggregator.go [hunk 19]
```diff
@@ -263,17 +269,27 @@ func (r *SyncCommitteeAggregatorRunner) ProcessPostConsensus(logger *zap.Logger,
 			}
 
 			submissionEnd := r.metrics.StartBeaconSubmission()
-
+			logger = logger.With(
```
Comment (y0sher) [inline protocol/v2/ssv/runner/sync_committee_aggregator.go:272]: move to different PR

File: protocol/v2/ssv/runner/validator_registration.go [hunk 20]
```diff
@@ -93,6 +93,7 @@ func (r *ValidatorRegistrationRunner) ProcessPreConsensus(logger *zap.Logger, si
 	}
 	specSig := phase0.BLSSignature{}
 	copy(specSig[:], fullSig)
+	r.metrics.EndPreConsensus()
```
Comment (y0sher) [inline protocol/v2/ssv/runner/validator_registration.go:96]: ditto

File: protocol/v2/ssv/runner/voluntary_exit.go [hunk 21]
```diff
@@ -89,21 +90,28 @@ func (r *VoluntaryExitRunner) ProcessPreConsensus(logger *zap.Logger, signedMsg
...
 
 	// create SignedVoluntaryExit using VoluntaryExit created on r.executeDuty() and reconstructed signature
 	signedVoluntaryExit := &phase0.SignedVoluntaryExit{
 		Message:   r.voluntaryExit,
 		Signature: specSig,
 	}
-
+	submissionTime := time.Now()
 	if err := r.beacon.SubmitVoluntaryExit(signedVoluntaryExit); err != nil {
+		logger.Error("failed to submit voluntary exit",
```
Comment (y0sher) [inline protocol/v2/ssv/runner/voluntary_exit.go:102]: ditto

File: protocol/v2/ssv/validator/non_committee_validator.go [hunk 22]
```diff
@@ -1,26 +1,51 @@
...
+	Network           specqbft.Network
+	Storage           *storage.QBFTStores
+	Operator          *spectypes.CommitteeMember
+	NetworkConfig     networkconfig.NetworkConfig
+	NewDecidedHandler qbftctrl.NewDecidedHandler
+	ValidatorStore    registrystorage.ValidatorStore
 }
 
-func NewNonCommitteeValidator(logger *zap.Logger, identifier spectypes.MessageID, opts Options) *NonCommitteeValidator {
+func NewNonCommitteeValidator(identifier exporter_message.MessageID, opts NonCommitteeOptions) *CommitteeObserver {
```
Comment (y0sher) [inline protocol/v2/ssv/validator/non_committee_validator.go:48]: the name `NewNonCommitteeValidator` not relevant anymore

File: protocol/v2/ssv/validator/non_committee_validator.go [hunk 23]
```diff
@@ -33,69 +58,93 @@ func NewNonCommitteeValidator(logger *zap.Logger, identifier spectypes.MessageID
...
 
-func (ncv *NonCommitteeValidator) ProcessMessage(logger *zap.Logger, msg *queue.DecodedSSVMessage) {
-	logger = logger.With(fields.MessageID(msg.MsgID), fields.Role(msg.MsgID.GetRoleType()))
+func (ncv *CommitteeObserver) ProcessMessage(msg *queue.DecodedSSVMessage) {
+	cid := spectypes.CommitteeID(msg.GetID().GetDutyExecutorID()[16:])
+	logger := ncv.logger.With(fields.CommitteeID(cid), fields.Role(msg.MsgID.GetRoleType()))
+
+	partialSigMessages := &spectypes.PartialSignatureMessages{}
+	if err := partialSigMessages.Decode(msg.SSVMessage.GetData()); err != nil {
+		logger.Debug("❗ failed to get partial signature message from network message", zap.Error(err))
```
Comment (y0sher) [inline protocol/v2/ssv/validator/non_committee_validator.go:82]: log should probably be Warn or Error

File: protocol/v2/ssv/validator/non_committee_validator.go [hunk 24]
```diff
@@ -33,69 +58,93 @@ func NewNonCommitteeValidator(logger *zap.Logger, identifier spectypes.MessageID
...
 
-		decMsg, err := specqbft.DecodeMessage(signedMsg.SSVMessage.Data)
-		if err != nil {
-			logger.Debug("❗ failed to decode message", zap.Error(err))
+	for key, quorum := range quorums {
+		role := ncv.getRole(msg, key.Root)
+		println("<<<<<<<<<<<<<<<<<<<<<<<<works>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")
+		println(role.String())
+		println("<<<<<<<<<<<<<<<<<<<<<<<<works>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")
+		validator := ncv.ValidatorStore.ValidatorByIndex(key.ValidatorIndex)
```
Comment (y0sher) [inline protocol/v2/ssv/validator/non_committee_validator.go:112]: remove logs here

File: protocol/v2/ssv/validator/non_committee_validator.go [hunk 25]
```diff
@@ -33,69 +58,93 @@ func NewNonCommitteeValidator(logger *zap.Logger, identifier spectypes.MessageID
...
-				}
-			}
+func (ncv *CommitteeObserver) getRole(msg *queue.DecodedSSVMessage, root [32]byte) exporter_message.RunnerRole {
+	if msg.MsgID.GetRoleType() == spectypes.RoleCommittee {
+		_, found := ncv.Roots[root]
+		if !found {
+			return exporter_message.RoleAttester
 		}
-		return
+		return exporter_message.RoleSyncCommittee
```
Comment (y0sher) [inline protocol/v2/ssv/validator/non_committee_validator.go:145]: this looks a bit hacky. so you save only sync commitee roots? but what if we have not root at all for this, we'll say its attester..

File: protocol/v2/ssv/validator/non_committee_validator.go [hunk 26]
```diff
@@ -106,3 +155,125 @@ func nonCommitteeInstanceContainerCapacity(fullNode bool) int {
...
+
+func (ncv *CommitteeObserver) processMessage(
+	signedMsg *spectypes.PartialSignatureMessages,
+) (map[validatorIndexAndRoot][]spectypes.OperatorID, error) {
+	quorums := make(map[validatorIndexAndRoot][]spectypes.OperatorID)
+
+	for _, msg := range signedMsg.Messages {
+		validator := ncv.ValidatorStore.ValidatorByIndex(msg.ValidatorIndex)
+		if validator == nil {
+			panic("fuck my life")
```
Comment (y0sher) [inline protocol/v2/ssv/validator/non_committee_validator.go:172]: I hope you hit this panic a few times, now please remove it :)

File: protocol/v2/ssv/validator/non_committee_validator.go [hunk 27]
```diff
@@ -106,3 +155,125 @@ func nonCommitteeInstanceContainerCapacity(fullNode bool) int {
...
+		ncv.logger.Debug("❗ failed to get beacon vote data", zap.Error(err))
+		return
+	}
+	cid := spectypes.CommitteeID(msg.GetID().GetDutyExecutorID()[16:])
+
+	ncv.logger.Info("✅ Got proposal message", fields.CommitteeID(cid))
+	ncv.Roots[beaconVote.BlockRoot] = spectypes.BNRoleSyncCommittee
+}
+
+//func (ncv *NonCommitteeValidator) onPostProposalMsg(msg *queue.DecodedSSVMessage) {
```
Comment (y0sher) [inline protocol/v2/ssv/validator/non_committee_validator.go:269]: remove comment

File: protocol/v2/ssv/validator/non_committee_validator.go [hunk 28]
```diff
@@ -106,3 +155,125 @@ func nonCommitteeInstanceContainerCapacity(fullNode bool) int {
...
+	// decode consensus data
+	beaconVote := &spectypes.BeaconVote{}
+	if err := beaconVote.Decode(msg.SignedSSVMessage.FullData); err != nil {
+		ncv.logger.Debug("❗ failed to get beacon vote data", zap.Error(err))
+		return
+	}
+	cid := spectypes.CommitteeID(msg.GetID().GetDutyExecutorID()[16:])
+
+	ncv.logger.Info("✅ Got proposal message", fields.CommitteeID(cid))
+	ncv.Roots[beaconVote.BlockRoot] = spectypes.BNRoleSyncCommittee
```
Comment (y0sher) [inline protocol/v2/ssv/validator/non_committee_validator.go:266]: how come its always sync committee?

**Review by y0sher — CHANGES_REQUESTED** [review]
In addition to my comments, please fix title and description for this PR.
good job :).

File: config/exporter.yaml (hunk 4 above)
Comment (guym-blox) [reply config/exporter.yaml:3]: it's not,
took it from e2e share 4 just for not having execption

File: protocol/v2/ssv/validator/non_committee_validator.go (hunk 28 above)
Comment (guym-blox) [reply protocol/v2/ssv/validator/non_committee_validator.go:266]: @y0sher because if the partialSig root is equal to beaconVote root it means it sync_committee

File: protocol/v2/ssv/validator/non_committee_validator.go (hunk 25 above)
Comment (guym-blox) [reply protocol/v2/ssv/validator/non_committee_validator.go:145]: @y0sher didn't understand

File: protocol/v2/ssv/runner/sync_committee.go (hunk 18 above)
Comment (guym-blox) [reply protocol/v2/ssv/runner/sync_committee.go:3]: ??

File: .gitlab-ci.yml [hunk 29]
```diff
@@ -83,7 +84,7 @@ Deploy nodes to hetzner stage:
     - .k8/hetzner-stage/scripts/deploy-cluster-65--68.sh $DOCKER_REPO_INFRA_STAGE $CI_COMMIT_SHA ssv $APP_REPLICAS_INFRA_STAGE hetzner.stage.k8s.local hetzner.stage.k8s.local stage.ssv.network $K8S_API_VERSION $STAGE_HEALTH_CHECK_IMAGE $SSV_NODES_CPU_LIMIT $SSV_NODES_MEM_LIMIT
     - .k8/hetzner-stage/scripts/deploy-cluster-69--72.sh $DOCKER_REPO_INFRA_STAGE $CI_COMMIT_SHA ssv $APP_REPLICAS_INFRA_STAGE hetzner.stage.k8s.local hetzner.stage.k8s.local stage.ssv.network $K8S_API_VERSION $STAGE_HEALTH_CHECK_IMAGE $SSV_NODES_CPU_LIMIT $SSV_NODES_MEM_LIMIT
   only:
-    - alan/no-fork
+    - stage
```
Comment (y0sher) [inline .gitlab-ci.yml:87]: alan/no-fork

File: .gitlab-ci.yml [hunk 30]
```diff
@@ -100,7 +101,7 @@ Deploy exporter to hetzner stage:
     - kubectl config get-contexts
     - .k8/hetzner-stage/scripts/deploy-holesky-exporters.sh $DOCKER_REPO_INFRA_STAGE $CI_COMMIT_SHA ssv $APP_REPLICAS_INFRA_STAGE hetzner.stage.k8s.local hetzner.stage.k8s.local stage.ssv.network $K8S_API_VERSION $SSV_EXPORTER_CPU_LIMIT $SSV_EXPORTER_MEM_LIMIT
   only:
-    - stage
+    - alan/no-fork
```
Comment (y0sher) [inline .gitlab-ci.yml:104]: alan/no-fork

File: exporter/api/msg.go [hunk 31]
```diff
@@ -113,7 +100,7 @@ type MessageFilter struct {
 	To uint64 `json:"to"`
 	// Role is the duty type, optional as it's relevant for IBFT data
 	Role string `json:"role,omitempty"`
-	// PublicKey is optional, used for fetching decided messages or information about specific validator/operator
+	// PublicKeys is optional, used for fetching decided messages or information about specific validator/operator
```
Comment (y0sher) [inline exporter/api/msg.go:103]:
```suggestion
	// PublicKey is optional, used for fetching decided messages or information about specific validator/operator
```

File: operator/validator/controller.go [hunk 32]
```diff
@@ -371,55 +374,81 @@ var nonCommitteeValidatorTTLs = map[spectypes.RunnerRole]phase0.Slot{
...
+	if err := pSigMessages.Decode(msg.SignedSSVMessage.SSVMessage.GetData()); err != nil {
 		return err
 	}
 
-	// Process the message.
-	defer ncv.Unlock()
-	ncv.ProcessMessage(c.logger, msg)
+	ncv.ProcessMessage(msg)
+	return nil
+}
```
Comment (y0sher) [inline operator/validator/controller.go:445]: You only need one function here, `handleNonCommitteeMessage` you can also lock and unlock the mutex outside of the function.

**Review by y0sher — CHANGES_REQUESTED** [review]
please elaborate a bit of more on the changes in the description

**Review by y0sher — APPROVED** [review]
